constraints.txt

```diff
@@ -7,58 +7,73 @@
 # need to submit a PR to pyca/cryptography, since they rely on it.
 
 alabaster==0.7.12
+async-generator==1.10
 attrs==21.2.0
 Babel==2.9.1
-backports.entry-points-selectable==1.1.1
-certifi==2021.10.8
-cffi==1.15.0
-charset-normalizer==2.0.7
-coverage==6.1.1
-cryptography==35.0.0
-distlib==0.3.3
-docutils==0.17.1
-filelock==3.3.2
-flake8==4.0.1
-idna==3.3
+backports.entry-points-selectable==1.1.0
+certifi==2021.5.30
+cffi==1.14.6
+charset-normalizer==2.0.4
+click==8.0.1
+coverage==5.5
+cryptography==3.4.8
+dataclasses-json==0.5.5
+distlib==0.3.2
+docutils==0.17
+filelock==3.0.12
+flake8==3.9.2
+idna==3.2
 imagesize==1.2.0
-importlib-metadata==4.2.0
```

Comment:
I think pinnings like this are still needed. See https://github.com/certbot/josepy/pull/122/checks#step:5:152.

What I personally have done when updating this file is to combine the output of `pip freeze` from the newest and oldest versions of Python we support. See https://github.com/certbot/josepy/issues/101 for more about this including considering switching to `poetry` here.

Reply:
Do not ask me twice for poetry ;)

In the mean time, I upgraded the contraint file using your method (Python 3.6 and Python 3.10).